src/core/remote/remote.ts

```diff
@@ -19,32 +34,134 @@ export default class Remote {
   }
 
   // used only by remote
-  send(message: RemoteRequest): Promise<any> {
+  send(message: ICreateRequest): Promise<any> {
     return request.register(message, this.sender);
   }
 
-  receiveMessage(message: string) {
-    const result = parse(message);
+  receiveMessage(data: string) {
+    const message = parse(data);
 
     if (this.isRemote()) {
-      return request.runCallback(result as RequestResult);
+      if (isEvent(message.type)) {
+        const { eventName, result } = message;
+
+        this.remote.emitEvent(eventName, result);
+        return;
+      }
+
+      return request.runCallback(message as IRequestResult);
+    }
+
+    // PROXY HANDLING
+    if (isEvent(message.type)) {
+      if (message.action === SUBSCRIPTION.ON) {
+        this.proxy.registerEvent(
+          message.clientId,
+          message.eventName,
+          (result: any) => {
+            this.sender({
+              from: XjsTypes.Proxy,
+              clientId: message.clientId,
+              type: EVENT,
+              eventName: message.eventName,
+              result,
+            });
+          }
+        );
+        return;
+      }
+
+      this.proxy.unregisterEvent(message.clientId, message.eventName);
+      return;
     }
 
-    // proxy
-    this.processRequest(result);
+    this.processRequest(message);
   }
 
-  async processRequest({ asyncId, fn, args }: ProxyRequest) {
+  async processRequest({ clientId, asyncId, fn, args }: IRequest) {
     const result = await this.exec(fn, ...args);
 
-    this.sender(
-      stringify({
-        asyncId,
-        result,
-      })
-    );
+    this.sender({
+      from: XjsTypes.Proxy,
+      clientId,
+      asyncId,
+      result,
+    });
   }
 
+  // EVENTS

```

Comment:
Is this going to be exposed to the consumer (ie. code that uses xjs framework)? If yes, then we'll have to make this a little more pretty... `remote.remote.emitEvent` isn't the prettiest of codes.

An alternative is to have these on their own classes and `remote.ts` would just use the correct class based on the remote type?

ex:
```
client.ts // the current remote.remote stuff
proxy.ts
remote.ts

// in remote.ts:
private events: Proxy | Client;

constructor({ type, exec }) {
  // ...current code
  
  if (type === XjsTypes.Proxy) {
    this.events = new Proxy({ exec });
  } else {
    this.events = new Client({ exec });
  }
}

// rest of code...

registerEvent: (obj) {
  return this.events.registerEvent(obj);
}
```

But if this is not intended to be used by the consumer, then we can leave this as it is :)